check-for-diff/action.yml

```
@@ -0,0 +1,14 @@
+name: Check for diff
+description: 'Check for changes between committed code and current state'
+
+inputs:
+  path_to_check:
+    description: 'Path to check'
+    required: false
+    default: ''
+
+runs:
+  using: "composite"
+  steps:
+    - run: git diff --color-words --exit-code && test -z "$(git status --porcelain ${{ inputs.path_to_check }})"
```

```suggestion
    - run: git diff ${{ inputs.path_to_check }} --color-words --exit-code && test -z "$(git status --porcelain ${{ inputs.path_to_check }})"
```
The first part with `git diff` must be narrowed to the path_to_check as well, right?
Probably the default value for `path_to_check` should be `'.'` then, right?

---

I'm not sure! It wasn't part of the original requirements, and I have already implemented the use of this action in all of our CI workflows and have encountered problems. It doesn't seem to be needed, at least at the moment. @zyv even raised the point whether in our APIs, we really need to restrict the git status check to a subfolder at all.

---

So from my point of view it does not make sense to narrow it down to a directory for the `git status --porcelain`, but not for `git diff`.

We want to find changes for a specific path or for the whole repository. `git status` shows you the changes, but it does not show you if there are additional files. That is why we have `git status --porcelain` additionally, but both commands should run on the same files.

Or am I missing something here?

The default value for `path_to_check` can still be empty string. That should work even after my suggested change. So I am okay with leaving that as it is.

---

I can't comment on how much sense the command makes in each specific context, but I can answer the following:

- `git diff --color-words --exit-code && test -z "$(git status --porcelain path/to/dir/)"`, as it is currently implemented in the composite action here, is used exactly like this for example in the profile API, [ci.yml#L44](https://github.com/moneymeets/django-profile-api/blob/78af2ade8e0d36ea114ec59c8c34309a55a2b392/.github/workflows/ci.yml#L44), and securities API [ci.yml#L46](https://github.com/moneymeets/django-securities-api/blob/ac44ba04a45d0070345681a4a7f388760bbdbe64/.github/workflows/ci.yml#L46)
- It is imaginable that one wants to A) check that committed code has not changed anywhere, and simultaneously B) check that there are no additional files in the path to check, but allow additional files elsewhere.

Does this help you at all, @pchorus? 

Since @zyv has also asked about this in his comment above, maybe @mm-matthias can clarify why, if at all, narrowing down the path to check is really needed in our APIs.

---

After discussion with @mm-matthias we agreed that both commands should run on the same files, so both of them should be narrowed to the given path, if it is provided. Therefore, please adopt the code as described above.

Furthermore, the `git status` command would be sufficient on its own and `git diff` does not find more changes, but we would like to keep it so that in the logs one can see immediately what has changed.

---

So did you also discuss why the restriction on the path is needed or not?

I'm file with making the action itself more flexible and allow it to be restricted if wanted, but actually I would like to avoid restrictions if possible.

---

The other thing I suggested is to actually define paths you want to exclude from check and otherwise always check everything.

The main question is what this action should do in the first place:
1. Make sure the files in a certain path didn't change (where path can also be the whole repository). One use case is to make sure the api is always regenerated from the spec.
2. Ensure the repository in general has no unexpected changes (except maybe some anticipated changes like .tmp folders being created by some tool or something like that)

In general 1) is a subset of 2). I think Yury is aiming to achieve 2) while the people in this PR seems to aim for 1).

I'd also prefer 2) if this is easy to implement. Maybe we can try to check the whole repository and even don't implement any exceptions first. My guess is that almost all of our repositories don't have any changes in whatever path there is.

---

I have checked whether repositories, that are using a git diff check, necessarily need a restriction to a specific directory:

- [x] [moneymeets-app](https://github.com/moneymeets/moneymeets-app/compare/feature/MD-6332-use-composite-action-to-check-for-diff) No restrictions used before
- [x] [moneymeets-dashboard](https://github.com/moneymeets/moneymeets-dashboard/compare/feature/MD-6332-use-composite-action-to-check-for-diff) Had restrictions, but also works without restrictions 
- [x] [django-insurance-api-v2](https://github.com/moneymeets/django-insurance-api-v2/compare/feature/MD-6332-use-composite-action-for-git-diff) Had restrictions, but also works without restrictions
- [x] [django-securities-api](https://github.com/moneymeets/django-securities-api/compare/feature/MD-6332-use-composite-action-for-git-diff) Had restrictions, but also works without restrictions
- [x] [django-profile-api](https://github.com/moneymeets/django-profile-api/compare/feature/MD-6332-use-composite-action-to-check-for-diff) Had restrictions, but also works without restrictions
- [x] [fondsnet-sdk](https://github.com/moneymeets/fondsnet-sdk/compare/feature/MD-6332-use-composite-action-to-check-for-diff) No restrictions used before
- [x] [morgen-und-morgen-sdk](https://github.com/moneymeets/morgen-und-morgen-sdk/compare/feature/MD-6332-use-composite-action-to-check-for-diff) Had restrictions, but also works without restrictions
- [x] [moneymeets-reporiting](https://github.com/moneymeets/moneymeets-reporting/compare/feature/MD-6332-use-composite-action-to-check-for-diff) No restrictions used before

In summary, currently, we do not need to restrict our diff checks to any subfolders. While this could become relevant in the future, I believe we can for now simply remove the option to restrict our checks. 

I have added a commit removing the parameter. As Pascal noted, `git diff` functionality is already covered by `git status`, but we want to keep it this as well to easily see what has changed in the logs. https://github.com/moneymeets/moneymeets-composite-actions/pull/4/commits/6e9008acfc739b4c06c511a05e4a4fedee99932b